### DISCUSSION TITLE
ATTiny3227: How to use TimerB for 1ms OVF Interrupt?

### DISCUSSION THREAD

**Discussion by Uwne** [SpenceKonde/megaTinyCore#591] [Q&A]
Hi,

I have a problem to read the datasheet and use the written instructions to set the the timerB for 1ms overflow interrupt.
As I also use millis() and want to use timerB(TCB1?). For an old project I use an attiny45 which I personally found much easier to 
set the timer. Can anyone give me hint how to set this up? The system is running at internal 8MHz.

Thank you.
Uwne

**Comment by grandaspanna** [comment]
I don't quite understand what you need it to do. Do you want Timer B to generate an interrupt every millisecond? Can you expand on your use-case a bit?

**Comment by Uwne** [comment]
Yes, I want to generate an interrupt every millisecond. 
Inside the isr a variable is counting up if a key is pressed (for example). If it has reached the given value (e.g. 610) it set´s another variable to high. This variable is used later in the program. In my opinion it is much more effective as the usage of millis().

But anyway, this is only one usage for it. I want to understand the way to set those registers in the new ATtinys.

**Comment by grandaspanna** [comment]
It's generally pretty straightforward. The datasheet for the parts gives a good overview of the different modes for each timer. I've used both TCA and the RTC to do exactly what you want (just run every millisecond to update something).

You normally just do some basic arithmetic to work out what your compare value needs to be for a given clock rate. Here's the setup for Timer A when running at 20MHz:

```
void setupTCA0_20(){

  // We are running at 20MHz, so configure the prescaler to 1024
  // This means the timer is running roughly at 20kHz
  // thus, it will take roughly 20 counts for 1ms to elapse
  TCA0.SINGLE.CNT = 20; 

  TCA0.SINGLE.INTCTRL = 1 << TCA_SINGLE_CMP0_bp   /* Compare 0 Interrupt: enabled */
                        | 0 << TCA_SINGLE_CMP1_bp /* Compare 1 Interrupt: disabled */
                        | 0 << TCA_SINGLE_CMP2_bp /* Compare 2 Interrupt: disabled */
                        | 0 << TCA_SINGLE_OVF_bp; /* Overflow Interrupt: disabled */

  TCA0.SINGLE.CTRLA = TCA_SINGLE_CLKSEL_DIV1024_gc /* System Clock / 1024 */
                      | 1 << TCA_SINGLE_ENABLE_bp /* Module Enable: enabled */;
}
```
Attach your code to the corresponding interrupt vector, so something like this:
```
ISR(TCA0_CMP0_vect)
{
  /* Insert your TCA Compare 0 Interrupt handling code here */
  current_millis = millis();
  
  /* The interrupt flag has to be cleared manually */
  TCA0.SINGLE.INTFLAGS = TCA_SINGLE_CMP0_bm;
}
```

Each timer type has its own cool features and the number of timers differs from part to part and check to see you're not stomping over whichever timer you've selected (or has been selected for you) for millis().

This code does something very similar for the RTC timer:
```
void RTC_0_init()
{
  // Set up the RTC to trigger an interrupt every 1ms
  
  while (RTC.STATUS > 0)
  {
    /* Wait for all register to be synchronized */
  }
  
  RTC.CTRLA = RTC_PRESCALER_DIV1_gc   /* 1 */
    | 1 << RTC_RTCEN_bp     /* Enable: enabled */            
    | 0 << RTC_RUNSTDBY_bp; /* Run In Standby: disabled */
    
  
  // RTC.CLKSEL = RTC_CLKSEL_INT32K_gc; /* 32KHz Internal Ultra Low Power Oscillator (OSCULP32K) */
    
  while (RTC.PITSTATUS > 0) { 
    /* Wait for all register to be synchronized */
    }
    
  RTC.PITCTRLA = RTC_PERIOD_CYC32_gc  /* RTC Clock Cycles 32 */
  | 1 << RTC_PITEN_bp; /* Enable: enabled */
  
  RTC.PITINTCTRL = 1 << RTC_PI_bp; /* Periodic Interrupt: enabled */
  
}
```

Corresponding interrupt setup:
```
ISR(RTC_PIT_vect)
{
  /* This code is called everytime the RTC PIT interrupt is triggered */
  current_millis = millis();
    
  /* The interrupt flag has to be cleared manually */
  RTC.PITINTFLAGS = RTC_PI_bm;
}
```

**Reply by Uwne** [reply]
Thank you very much for the detailed and quick reply.
I will try it out and let you know my result asap!

**Comment by SpenceKonde** [comment]
I'm curious as to the use case, since there is already a timer with an ISR incrementing a counter if milis isn't disabled. Why can this functionality you need not be achiveved just using millis?

**Reply by grandaspanna** [reply]
I'll profess ignorance here - what is this ISR and how do I piggy back off it? Might help one of my scenarios.

**Reply by Uwne** [reply]
Happy new year to all of you.
I don´t want to use millis as of the rollover problem I have read every ~50days. 
Is there a way to reset the millis timer to prevent the rollover issue?

I tried to use the subroutine from the first post with the setupTCA0_20() but without success.
As it was my first try to use it I cannot say what I am doing wrong. Actually my time is limited this week,
as start of next week I have some more time and also can post my program / changes.

**Reply by hmeijdam** [reply]
Now I am getting curious why the rollover is perceived as a problem for you.
Did you read:   https://www.gammon.com.au/millis

**Reply by SpenceKonde** [reply]
You can't piggy back on an ISR; there can only be one definition of a given ISR in one program. . But it sounds like the OP just needs a number that increments once per millisecond. Which is exactly what millis does, and it seems like it would be redundant (not to mention blocking up the other timer that would be used for servo, tone, or a number of other libraries) to have a second timer counting milliseconds....

The millis rollover isn't a problem unless you mishandle the numbers it returns when doing math with it. 

You do all the math as unsigned longs and subtract the old time from the new time, and the result is the difference,.
```
//in global scope
unsigned long oldtime;
// somewhere in program
oldtime=millis();
//somewhere else
unsigned long time_passed = millis() - oldtime;
// this will always work except for cases where you're trying to measure times of over 50 days to the millisecond, in which case you have an X-Y problem. 
if (millis() - oldtime > 1000) { ... }
// This will also work. 
// But below, using addition....
if (millis() > (oldtime + 1000)) { ... } 
// this will break at the rollover; don't do that.
```
Another misuse of millis 
```
unsigned long do_something_at;
//...
do_someting_at = millis() + 30000;
//...
if (millis() > do_something_at) 
// will break around the rollover. 
```
The rules are pretty simple:
1. Store the first millis() timestamp. Don't modify it now. 
2. Subtract the earlier timestamp from a later one to find how much time has elapsed. 
3. Do whatever you want with that difference.

**Comment by Uwne** [comment]
Hi,

yes I am sorry, now it is really clear. I do not need to measure times for more than 50 days, just a few seconds.
I thought the rollover would be (always) a problem, which it isn´t.
I just make a test programm where I came across a problem while using millis():

```
void loop() 
  {    
    unsigned long currentMillis = millis(); 
   
    // Test with Relais
    
    if (digitalRead(button1Pin) == LOW)       
      {
        digitalWrite(relaisPin, LOW); Serial.println("Relais ON");          
         
        if (currentMillis - previousMillis2 >= interval) //Interval = 2000
          {                 
            Serial.print(currentMillis);Serial.print("\t");Serial.print(previousMillis2);Serial.print("\t");Serial.println(currentMillis-previousMillis2);
            previousMillis2 = currentMillis;                                      
            digitalWrite(relaisPin, HIGH); Serial.println("Relais OFF");      
          }       
      }         
  }
```
The serial prints are only for debugging.
If I use this code the relaisPin will get LOW and directly HIGH after pressing the button1. 
It has to because on power up previousMillis2 is 0 and will only be updated after previousMillis2 = currenMillis.

What do I want to do?
After I switch a button to ON I want to start a subroutine which should run for time <interval>.
I cannot see my mistake or the solution actually.

**Reply by hmeijdam** [reply]
Maybe because you are not [debouncing](https://arduinogetstarted.com/tutorials/arduino-button-debounce) the button.

[edit] looking at it again, you only continue to check for the interval if the button remains pressed. As soon as you release the button the check against interval will never be done.

**Reply by Uwne** [reply]
Yes that is correct, the button must be hold pressed to make it count.
This `if (currentMillis - previousMillis2 >= interval)` will always be truth on the first time.

I need to find another way to count the millis() how long a button is pressed.
